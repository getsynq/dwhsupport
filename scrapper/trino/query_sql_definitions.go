package trino

import (
	"context"
	_ "embed"
	"strings"

	dwhexec "github.com/getsynq/dwhsupport/exec"
	"github.com/getsynq/dwhsupport/exec/stdsql"
	"github.com/getsynq/dwhsupport/scrapper"
)

//go:embed query_sql_definitions.sql
var querySqlDefinitionsSQL string

func (e *TrinoScrapper) QuerySqlDefinitions(ctx context.Context) ([]*scrapper.SqlDefinitionRow, error) {
	query := querySqlDefinitionsSQL
	db := e.executor.GetDb()
	var out []*scrapper.SqlDefinitionRow

	availableCatalogs, err := e.allAvailableCatalogs.Get()
	if err != nil {
		return nil, err
	}

	for _, catalog := range availableCatalogs {
		if !catalog.IsAccepted {
			continue
		}
		catalogQuery := strings.Replace(query, "{{catalog}}", catalog.CatalogName, -1)
		res, err := stdsql.QueryMany(ctx, db, catalogQuery,
			dwhexec.WithPostProcessors(func(row *scrapper.SqlDefinitionRow) (*scrapper.SqlDefinitionRow, error) {
				row.Instance = e.conf.Host
				return row, nil
			}),
		)
		if err != nil {
			return nil, err
		}
		out = append(out, res...)
	}
	return out, nil
}
