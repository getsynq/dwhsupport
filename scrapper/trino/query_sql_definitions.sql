with tables as (
    select
        table_catalog as database,
        table_schema as schema,
        table_name,
        table_type
    from {{catalog}}.information_schema.tables
    where table_schema not in ('information_schema')
)
select
    t.database,
    t.schema,
    t.table_name as "table",
    mv.name is not null OR t.table_type = 'VIEW' as is_view,
    coalesce(mv.definition, v.view_definition) as sql
from tables t
left join {{catalog}}.information_schema.views v
    on t.schema = v.table_schema and t.table_name = v.table_name
LEFT JOIN system.metadata.materialized_views mv
    ON t.database = mv.catalog_name
    AND t.schema = mv.schema_name
    AND t.table_name = mv.name
where mv.definition is not null OR v.view_definition is not null
order by t.database, t.schema, t.table_name 