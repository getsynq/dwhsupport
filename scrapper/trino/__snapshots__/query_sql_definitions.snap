
[TestQuerySnapshotGeneration/AllFeaturesDisabled - 1]
with tables as (
    select
        table_catalog as database,
        table_schema as schema,
        table_name,
        table_type
    from test_catalog.information_schema.tables
    where table_schema not in ('information_schema')
)
select
    t.database,
    t.schema,
    t.table_name as "table",
    (t.table_type = 'VIEW' AND v.view_definition IS NOT NULL) as is_view,
    false as is_materialized_view,
    coalesce(v.view_definition, '') as sql
from tables t
left join test_catalog.information_schema.views v
    on t.schema = v.table_schema and t.table_name = v.table_name

order by t.database, t.schema, t.table_name
---

[TestQuerySnapshotGeneration/AllFeaturesEnabled - 1]
with tables as (
    select
        table_catalog as database,
        table_schema as schema,
        table_name,
        table_type
    from test_catalog.information_schema.tables
    where table_schema not in ('information_schema')
)
select
    t.database,
    t.schema,
    t.table_name as "table",
    (t.table_type = 'VIEW' AND v.view_definition IS NOT NULL) as is_view,
    (mv.name IS NOT NULL) as is_materialized_view,
    coalesce(mv.definition, v.view_definition, '') as sql
from tables t
left join test_catalog.information_schema.views v
    on t.schema = v.table_schema and t.table_name = v.table_name
LEFT JOIN system.metadata.materialized_views mv ON t.database = mv.catalog_name AND t.schema = mv.schema_name AND t.table_name = mv.name
order by t.database, t.schema, t.table_name
---

[TestQuerySnapshotGeneration/OnlyMaterializedViews - 1]
with tables as (
    select
        table_catalog as database,
        table_schema as schema,
        table_name,
        table_type
    from test_catalog.information_schema.tables
    where table_schema not in ('information_schema')
)
select
    t.database,
    t.schema,
    t.table_name as "table",
    (t.table_type = 'VIEW' AND v.view_definition IS NOT NULL) as is_view,
    (mv.name IS NOT NULL) as is_materialized_view,
    coalesce(mv.definition, v.view_definition, '') as sql
from tables t
left join test_catalog.information_schema.views v
    on t.schema = v.table_schema and t.table_name = v.table_name
LEFT JOIN system.metadata.materialized_views mv ON t.database = mv.catalog_name AND t.schema = mv.schema_name AND t.table_name = mv.name
order by t.database, t.schema, t.table_name
---

[TestQuerySnapshotGeneration/OnlyTableComments - 1]
with tables as (
    select
        table_catalog as database,
        table_schema as schema,
        table_name,
        table_type
    from test_catalog.information_schema.tables
    where table_schema not in ('information_schema')
)
select
    t.database,
    t.schema,
    t.table_name as "table",
    (t.table_type = 'VIEW' AND v.view_definition IS NOT NULL) as is_view,
    false as is_materialized_view,
    coalesce(v.view_definition, '') as sql
from tables t
left join test_catalog.information_schema.views v
    on t.schema = v.table_schema and t.table_name = v.table_name

order by t.database, t.schema, t.table_name
---
